pkg/node/manager/manager.go

```diff
@@ -332,134 +333,138 @@ func (m *manager) legacyNodeIpBehavior() bool {
 	return true
 }
 
+func (m *manager) nodeAddressHasTunnelIP(address nodeTypes.Address) bool {
+	// If the host firewall is enabled, all traffic to remote nodes must go
+	// through the tunnel to preserve the source identity as part of the
+	// encapsulation. In encryption case we also want to use vxlan device
+	// to create symmetric traffic when sending nodeIP->pod and pod->nodeIP.
+	return address.Type == addressing.NodeCiliumInternalIP || m.conf.EncryptionEnabled() ||
+		option.Config.EnableHostFirewall || option.Config.JoinCluster
+}
+
+func (m *manager) nodeAddressHasEncryptKey(address nodeTypes.Address) bool {
+	return (m.conf.NodeEncryptionEnabled() ||
+		// If we are doing encryption, but not node based encryption, then do not
+		// add a key to the nodeIPs so that we avoid a trip through stack and attempting
+		// to encrypt something we know does not have an encryption policy installed
+		// in the datapath. By setting key=0 and tunnelIP this will result in traffic
+		// being sent unencrypted over overlay device.
+		(address.Type != addressing.NodeExternalIP && address.Type != addressing.NodeInternalIP)) &&
+		// Also ignore any remote node's key if the local node opted to not perform
+		// node-to-node encryption
+		!node.GetOptOutNodeEncryption()
+}
+
+func (m *manager) nodeAddressSkipsIPCache(address nodeTypes.Address) bool {
+	return m.legacyNodeIpBehavior() && address.Type != addressing.NodeCiliumInternalIP
+}
+
+func nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels, hasOverride bool) {
```

Comment:
I think this should maybe take into account `m.conf.RemoteNodeIdentitiesEnabled` like the earlier corresponding code did? Otherwise even with remote node identities disabled, it will set the labels to remote node and break connectivity for users with that setting.

Reply:
Unfortunately I don't think the test coverage is likely to tell you much on this front :/ I just noticed the difference while comparing this patch just a few lines down in the next hunk (the code that got removed).

Reply:
Good catch, addressed:

<details><summary>Details</summary>
<p>

```diff
commit 40ac0845b550f99fbfca5632b8d26541d5f0dcc3
Author: Chris Tarazi <chris@isovalent.com>
Date:   Tue Apr 11 22:03:28 2023 -0700

    fixup! nodemanager: Move to async IPCache API
    
    Signed-off-by: Chris Tarazi <chris@isovalent.com>

diff --git a/pkg/node/manager/manager.go b/pkg/node/manager/manager.go
index c341e8fdcb..f2d6ae18cc 100644
--- a/pkg/node/manager/manager.go
+++ b/pkg/node/manager/manager.go
@@ -359,14 +359,18 @@ func (m *manager) nodeAddressSkipsIPCache(address nodeTypes.Address) bool {
 	return m.legacyNodeIpBehavior() && address.Type != addressing.NodeCiliumInternalIP
 }
 
-func nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels, hasOverride bool) {
+func (m *manager) nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels, hasOverride bool) {
 	nodeLabels = labels.NewFrom(labels.LabelRemoteNode)
-	if n.IsLocal() {
+	if m.conf.RemoteNodeIdentitiesEnabled() {
+		if n.IsLocal() {
+			nodeLabels = labels.NewFrom(labels.LabelHost)
+		} else if !identity.NumericIdentity(n.NodeIdentity).IsReservedIdentity() {
+			// This needs to match clustermesh-apiserver's VMManager.AllocateNodeIdentity
+			nodeLabels = labels.Map2Labels(n.Labels, labels.LabelSourceK8s)
+			hasOverride = true
+		}
+	} else {
 		nodeLabels = labels.NewFrom(labels.LabelHost)
-	} else if !identity.NumericIdentity(n.NodeIdentity).IsReservedIdentity() {
-		// This needs to match clustermesh-apiserver's VMManager.AllocateNodeIdentity
-		nodeLabels = labels.Map2Labels(n.Labels, labels.LabelSourceK8s)
-		hasOverride = true
 	}
 	return nodeLabels, hasOverride
 }
@@ -390,7 +394,7 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
 	}
 
 	resource := ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindNode, "", n.Name)
-	nodeLabels, nodeIdentityOverride := nodeIdentityLabels(n)
+	nodeLabels, nodeIdentityOverride := m.nodeIdentityLabels(n)
 
 	var nodeIPsAdded, healthIPsAdded, ingressIPsAdded []netip.Prefix
 
@@ -519,7 +523,7 @@ func (m *manager) removeNodeFromIPCache(oldNode nodeTypes.Node, resource ipcache
 		// See comment in NodeUpdated().
 		oldNodeIP, _ = ip.AddrFromIP(nIP)
 	}
-	oldNodeLabels, oldNodeIdentityOverride := nodeIdentityLabels(oldNode)
+	oldNodeLabels, oldNodeIdentityOverride := m.nodeIdentityLabels(oldNode)
 
 	// Delete the old node IP addresses if they have changed in this node.
 	for _, address := range oldNode.IPAddresses {

```

</p>
</details>